### PR TITLE
Aggregate coffee calls on backend

### PR DESCRIPTION
**Pull request by t-animal**

### CODE REVIEW

File: backend/src/message-pipeline.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+class AggregatedCoffeeCall implements CoffeeCall {
+  status: CoffeeCallStatus = 'inactive';
+  participants: string[] = [];
+
+  constructor(public messages: TimedCommand[]) {
+    if (messages.length === 0) {
+      return;
+    }
+
+    const participantsTmp = new Set<string>();
```
Comment (yogan) [inline backend/src/message-pipeline.ts]: dropping the `Tmp` is fine (`participants` vs. `this.participants`)

File: backend/src/main.ts [hunk 2]
```diff
@@ -1,27 +1,49 @@
-import * as dotenv from "dotenv";
-dotenv.config();
-
 import { createServer } from "http";
-import { Server } from "socket.io";
+import { Server, Socket } from "socket.io";
+import { shareReplay, Subject, take, timeout, } from "rxjs";
```
Comment (yogan) [inline backend/src/main.ts:3]:
```suggestion
import { Server } from "socket.io";
import { shareReplay, Subject } from "rxjs";
```

File: frontend-web/src/features/call-for-coffee/coffee-call-state/coffee-call-state.tsx [hunk 3]
```diff
@@ -1,18 +1,18 @@
 import { Accessor, Component, For, Show, splitProps } from "solid-js";
-import { CoffeeCall, CoffeeCallStatus } from "../backend-communication/call-aggregation";
+import type { CoffeeCall } from "../backend-communication/models";
 
 export const CoffeeCallState: Component<{ coffeeCall: CoffeeCall }> = (props) => {
-  const participantsAsList = () => Array.from(props.coffeeCall.participants.values()).join(',');
+  const participantsAsList = () => props.coffeeCall.participants.join(',');
```
Comment (yogan) [inline frontend-web/src/features/call-for-coffee/coffee-call-state/coffee-call-state.tsx]:
```suggestion
  const participantsAsList = () => props.coffeeCall.participants.join(', ');
```

Not relevant to the PR, and we might at some point render the participants differently, but it's a easy fix for now.

File: backend/src/main.ts [hunk 4]
```diff
@@ -1,27 +1,49 @@
...
 });
 
+const messageSubject = new Subject<Command>();
+const processedCoffeeCalls = messageSubject.pipe(convertCommandsToCalls, shareReplay(1));
+
+function broadcastCommand(command: Command) {
+    const timed = { ...command, broadcastAt: new Date() };
+    console.log('→ broadcast command msg:', timed);
+    io.sockets.emit("coffee", timed);
+}
```
Comment (yogan) [inline backend/src/main.ts]: With the introduction of aggregated calls in the BE we could basically just remove this.  
Besides code reduction, it'll also make it easier to analyze the protocol by observing the messages sent back and forth (less noise on the line).

File: backend/src/main.ts [hunk 5]
```diff
@@ -1,27 +1,49 @@
...
-            ...JSON.parse(data),
-            broadcastAt: new Date().toISOString()
-        }));
-    });
+
+    socket.on("coffee", logCommand);
+    socket.on("coffee", broadcastCommand)
+    socket.on("coffee", putCommandIntoPipeline);
+
+    processedCoffeeCalls.subscribe((coffeeCall) => socket.emit("coffeeCalls", coffeeCall));
```
Comment (yogan) [inline backend/src/main.ts]: Suggestion to rename the message channels (events?):
- `coffee` → `coffeeRequest` (coming from clients to the BE: `join`, `leave`, `start`)
- `coffeeCall` keep as it is (sent from BE to clients whenever something changes)

**Review by yogan — COMMENTED** [review]
Great work, I like the approach to aggregate on the BE and only send out accumulated calls.  
Removes the necessity to implement that logic in each and every client (in potentially different ways).

Since this is a protocol change, I think it is a good point in time to re-consider the event names that we are using for messages sent via Socket.IO. I've added a [suggestion for a possible name change](https://github.com/erlangen-dev/coffee-caller/pull/33#discussion_r978483021).

File: backend/src/main.ts (hunk 2 above)
Comment (t-animal) [reply backend/src/main.ts:3]: Will be fixed by the formatting PR across all files

**Pull request merged**